Add Fossil VCS support

demize opened the pull request:
Started using Fossil for a project, and was annoyed that I didn't have status in my prompt. This commit adds it in!

I will admit I'm not overly familiar with Fossil, so this may be less than ideal (and might be missing important statuses), but in my testing it works pretty well. I added a new option to enable Fossil support (disabled by default, because most people won't need it; Fossil is pretty quick, so it shouldn't have the same issues as Mercurial), updated the README, and tried to stick to the same coding style as the existing code.

My one complaint about my own work is the default (non-NF) Fossil glyph, which is just the bone emoji. The NF bone emoji is alright (though I'd prefer something *more* evocative of "fossil") but I couldn't think of a better symbol outside NF.

bobthecow commented on functions/__bobthefish_glyphs.fish:
```diff
@@ -83,6 +86,7 @@ function __bobthefish_glyphs -S -d 'Define glyphs used by bobthefish'
     set git_stashed_glyph    \uF0C6 '' # nf-fa-paperclip
     set git_untracked_glyph  \uF128 '' # nf-fa-question
     # set git_untracked_glyph  \uF141 '' # nf-fa-ellipsis_h
+    set fossil_glyph '󰂹'
```
for nerd font glyphs, please use the escaped unicode code point rather than the literal character, and add the NF name as a comment. this makes it much easier to update when things change :)

demize replied:
Used the escaped code point here, and for the bone emoji too :)

bobthecow commented on functions/__bobthefish_glyphs.fish:
```diff
@@ -49,6 +49,9 @@ function __bobthefish_glyphs -S -d 'Define glyphs used by bobthefish'
   set -x git_plus_minus_glyph '±'
 
+  # Fossil glyph (it reuses most of the git glyphs)
+  set -x fossil_glyph '🦴'
```
this should probably be suppressed in the `theme_avoid_ambiguous_glyphs` block below. that's used for terminals that don't love modern characters and can't deal with things like emojis. i'm not sure there's a good non-emoji fallback, so maybe just have it be empty string?

demize replied:
Tried both `''` and `' '`, went with the null string because it looks better.

bobthecow commented on functions/fish_prompt.fish:
```diff
@@ -1020,6 +1037,54 @@ end
 # ==============================
 
+function __bobthefish_prompt_fossil -S -a fossil_root_dir -a real_pwd -d 'Display the actual fossil state'
+    set fossil_statuses (command fossil changes --differ 2>/dev/null | cut -d' ' -f1 | sort -u)
```
prolly should use `set -l` or `-f` for all of these, to avoid contaminating user variables from outside the function scope.

bobthecow commented on functions/fish_prompt.fish:
```diff
@@ -1020,6 +1037,54 @@ end
+    __bobthefish_start_segment $flag_colors
+    echo -ns $fossil_glyph ' '
+
+    __bobthefish_start_segment $flag_colors
```
do you want to start a new segment or just prefix the current segment with a bone? it's probably more bobthefishy to keep it in a single segment?

demize replied:
Good point; much better as a single segment.

bobthecow commented on functions/fish_prompt.fish:
```diff
@@ -1255,6 +1321,8 @@ function fish_prompt -d 'bobthefish, a fish theme optimized for awesome'
     else if [ "$hg_root_dir" ]
         __bobthefish_prompt_hg $hg_root_dir $real_pwd
+    else if [ "$fossil_root_dir" ]
+        __bobthefish_prompt_fossil $fossil_root_dir $real_pwd
```
i dunno if there's a better way to do this, but for git+hg i added a naive nested repo support. doing this three ways feels like it'd be a mess.

maybe we could do something with `string length` of the various paths. or maybe it's not even worth worrying about.

demize replied:
```fish
switch (path sort -r "$git_root_dir" "$hg_root_dir" "$fossil_root_dir")[1]
  case "$git_root_dir"
    __bobthefish_prompt_git $git_root_dir $real_pwd
  case "$hg_root_dir"
    __bobthefish_prompt_hg $hg_root_dir $real_pwd
  case "$fossil_root_dir"
    __bobthefish_prompt_fossil $fossil_root_dir $real_pwd
end
```

I think this would work, and is probably least messy. Not sure if it's worth it either, but I think this is an elegant enough solution. And it could replace the entire series of ifs/elses, since `path sort` will handle empty paths as if they're above `/`:

```shell
$ path sort -r "git" "" "git/fossil repo"
git/fossil repo
git

$ # previous line intentionally left blank
```

bobthecow replied:
Yeah I don't hate this approach. It's definitely easier to follow than the nested ifs and switches.

Threw it on the old `--profile` and it looks like `path sort` for two VCS paths is about a 3x penalty vs ifs and cases and string matching, and a 2x penalty when comparing three paths. I think that's something we can afford.

Let's do this.

bobthecow replied:
Note that the "no VCS directory at all" case has to be handled explicitly, otherwise `path sort` will return an empty string, and it'll match the first `case "$git_root_dir"` (which itself is an empty string too). So something like this:

```fish
switch (path sort -r "$git_root_dir" "$hg_root_dir" "$fossil_root_dir")[1]
    case ''
        __bobthefish_prompt_dir $real_pwd
    case "$git_root_dir"
        __bobthefish_prompt_git $git_root_dir $real_pwd
    case "$hg_root_dir"
        __bobthefish_prompt_hg $hg_root_dir $real_pwd
    case "$fossil_root_dir"
        __bobthefish_prompt_fossil $fossil_root_dir $real_pwd
end
```

demize commented:
Addressed all the feedback, tweaked the prompt a little bit more, and also noticed that sometimes Fossil would take 1-2 seconds to print the `fossil info` output. Tracking that down and coming up with a solution took me longer than anticipated, but I got something pretty good working with `fossil json status` I think!

bobthecow commented on functions/fish_prompt.fish:
```diff
@@ -197,14 +197,15 @@ end
     and return
 
-    command fossil info 2>/dev/null | sed '3q;d' | string split ' ' -f2 -n | string trim --right --chars=/
+    set -f dir (command fossil json status 2>/dev/null | grep localRoot | string split ':' -f2 | string trim --chars='"/,')
```
rather than calling it a second time, should we save the output from the start of the function and use that?

demize replied:
Yeah, that makes more sense than what I was doing... just inlined the `set` into the first check, which works nicely.

bobthecow commented:
Awesome, thanks for this!
